Add XLA-only merge that can merge all types.

bas-aarts opened the pull request:
This prevents insertion of H2D and D2H copies when XLA-GPU clusters
have int32 outputs. This merge is only used the merge the outputs
from the XlaRun and the the PartitionedCall node.

Commit:
Add XLA-only merge that can merge all types.

This prevents insertion of H2D and D2H copies when XLA-GPU clusters
have int32 outputs. This merge is only used the merge the outputs
from the XlaRun and the the PartitionedCall node.

sanjoy commented on tensorflow/compiler/jit/kernels/xla_ops.h:
```diff
@@ -175,6 +175,16 @@ class XlaRunOp : public OpKernel {
+  void Compute(OpKernelContext* ctx) override;
+
+ private:
+  const XlaPlatformInfo platform_info_;
```
Why do you need this?

bas-aarts replied:
no. will remove it.

sanjoy commented on tensorflow/compiler/jit/kernels/xla_ops.cc:
```diff
@@ -648,6 +659,12 @@ REGISTER_KERNEL_BUILDER(Name("_XlaCompile")
 
 REGISTER_KERNEL_BUILDER(Name("_XlaRun").Device(DEVICE_CPU), XlaRunOp);
-REGISTER_KERNEL_BUILDER(Name("_XlaRun").Device(DEVICE_GPU), XlaRunOp);
+REGISTER_KERNEL_BUILDER(Name("_XlaRun")
```
Wow!  Was this working because TF puts all strings on the host?

bas-aarts replied:
That would be my guess, but did not investigate
Just noticed it will making this change, and fixed it.

sanjoy commented on tensorflow/compiler/jit/build_xla_ops_pass.cc:
```diff
@@ -135,9 +135,9 @@ void MergeOutgoingDataEdges(const Scope& s, Node* old_node, Node* new_node,
-      ops::Merge merge_op(s.WithOpName("merge_oidx_", oidx),
-                          {Output(old_node, oidx), new_output});
-      merged_output = merged_outputs[oidx] = merge_op.output;
+      ops::_XlaMerge xla_merge_op(s.WithOpName("merge_oidx_", oidx),
```
Does this work for `DT_RESOURCE` outputs?

bas-aarts replied:
It should. haven't seen any issue locally.

sanjoy replied:
Looks like both strings and resources are handled specially and automatically placed on the host: https://github.com/tensorflow/tensorflow/blob/cca6113667e10b1f3405dbe3b1bd3d983f9cf420/tensorflow/core/framework/memory_types.cc#L146

sanjoy commented on tensorflow/compiler/jit/kernels/xla_ops.h:
```diff
@@ -175,6 +175,16 @@ class XlaRunOp : public OpKernel {
   const XlaPlatformInfo platform_info_;
 };
 
+class XlaMergeOp : public OpKernel {
```
Can you just use `tensorflow::MergeOp` instead of creating a new implementation?

bas-aarts replied:
I initially tried to use the same implementation. The allocation for the value_index scalar (through context->allocate_output(...) doesn't work when called through XLA. I then tried an implementation using GetAllocator(..., patform_info_) which did work. I subsequently decided to limit the _XlaMerge op to not even have a value index. I didn't see a need for it, and build_xla_ops_pass.cc doesn't use it either.
(When removing the index, I forgot to remove platform_info_, which you commented about earlier).

sanjoy replied:
Ok.  Please add a comment on the `.Doc` for `_XlaMerge` that it intentionally does not have the `value_index` output.

bas-aarts replied:
done

sanjoy commented on tensorflow/compiler/jit/xla_device_ops.h:
```diff
@@ -72,6 +72,11 @@ class XlaAssignVariableOp : public OpKernel {
 #define REGISTER_XLA_RUN_KERNEL(DEVICE, KERNEL, TYPES) \
   REGISTER_KERNEL_BUILDER(Name("_XlaRun").Device(DEVICE), KERNEL);
 
+#define REGISTER_XLA_MERGE_KERNEL(DEVICE, KERNEL, TYPES)  \
```
When do you need this?  Since XLA_* device clusters are always compiled, we should not need to generate merge ops for them.

bas-aarts replied:
I'm not sure.
But how is op different from _XlaRun and _XlaCompile in this context?

sanjoy replied:
For ops placed on XLA_* devices we still create clusters as usual.  The semantic for XLA_* devices is that:

  1. Ops placed on these devices must be compiled
  2. Ops placed on these devices will be opportunistically clustered

Since ops placed on these devices **must** be compiled we do not need (rather cannot have) the lazy compilation scheme so we should be generating just the `_XlaCompile` and `_XlaRun` nodes placed on XLA_* devices.  Now that I think of it, it might be more logical to place these `_XlaCompile` and `_XlaRun` nodes on `GPU` and not `XLA_GPU`, but that's a separate change.

I also want to stress that the `XLA_*` devices are only there for unit tests, we usually push back on users trying to use them in regular code.

bas-aarts commented on tensorflow/compiler/jit/ops/xla_ops.cc:
```diff
@@ -95,4 +95,18 @@ Executes a TensorFlow function previously compiled into a LocalExecutable by an
+      c->set_output(0, c->input(0));
+      return Status::OK();
+    })
+    .Doc(R"(XLA Merge Op. For use by the XLA JIT only.
```
will add more info to the .Doc string

Commit:
address comments on commit e4c76ebc7790368227f2aa4695c177cff364e463

Add XLA-only merge that can merge all types.

This prevents insertion of H2D and D2H copies when XLA-GPU clusters
have int32 outputs. This merge is only used the merge the outputs
from the XlaRun and the the PartitionedCall node.

sanjoy commented on tensorflow/compiler/jit/ops/xla_ops.cc:
```diff
@@ -95,4 +95,22 @@ Executes a TensorFlow function previously compiled into a LocalExecutable by an
+    .Doc(R"(XLA Merge Op. For use by the XLA JIT only.
+
+Merges the outputs from the PartitionedCall node and the _XlaRun node.
+Unlike the TensorFlow merge op, _XlaMerge supports merging inputs of all types.
```
This sounds a bit misleading.  Even `Merge` supports all data types; the difference is that Merge places int32 types on the host whereas `_XlaMerge` places them on the device.

bas-aarts replied:
agreed. will clarify

Commit:
address comments on commit d628b34f78f291692c8b4ef0da54f1b91f24291d

Add XLA-only merge that can merge all types.

This prevents insertion of H2D and D2H copies when XLA-GPU clusters
have int32 outputs. This merge is only used the merge the outputs
from the XlaRun and the the PartitionedCall node.

sanjoy commented:
LGTM from me, but I want @rmlarsen to also take a look.

rmlarsen reviewed (approved):
Short of fixing the int32 situation entirely, I'm not sure there is a less hacky way to do this, so I'll approve to unblock these performance improvements.

sanjoy commented:
Update: I ran the internal benchmarks with this and found some minor regressions, checking if they are noise or not.

bas-aarts commented:
> Update: I ran the internal benchmarks with this and found some minor regressions, checking if they are noise or not.

Checking in too see if any real issues were discovered.

sanjoy commented:
> Checking in too see if any real issues were discovered.

No issues, the only "regression" was a noisy benchmark.  I've kicked off the submission internally.
